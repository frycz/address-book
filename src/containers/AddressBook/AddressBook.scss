.address-book {
    &__top-bar {
        display: flex;
        flex-direction: row;
        justify-content: center;
        align-items: center;

        position: sticky;
        top: 0;
        background-color: #ffffff;
    }

    &__title {
        flex: 1;
        text-align: center;
    }

    &__message {
        text-align: center;
    }

    &__details-modal {
        position: absolute;
        top: 50%;
        left: 50%;
        right: 50%;
        bottom: 50%;
        width: 400px;
        height: 250px;
        transform: translate(-50%, -50%);
        border-radius: 5px;
        border: 1px solid gray;
        padding: 10px 20px;
        box-sizing: border-box;
        background-color: white;

        &:focus {
            outline: none;
        }
    }

    &__details-container {
        display: flex;
        justify-content: space-around;
    }
}
