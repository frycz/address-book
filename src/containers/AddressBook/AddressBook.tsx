import React, { useEffect } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import Modal from "react-modal";
import { useBottomScrollListener } from "react-bottom-scroll-listener";
import { State, User, Countries } from "../../redux/store";
import { getUsers, displayPage } from "../../redux/book/actions";
import { paths } from "../../router";
import "./AddressBook.scss";

export interface Props {
  users: User[][];
  maxPage: number;
  isFetching: boolean;
  isError: boolean;
  countries: Countries;
  displayedPage: number;
  getUsers: typeof getUsers;
  displayPage: typeof displayPage;
}

/**
 * Main component for displaying address book
 * 
 * @param {Props} props 
 */
export const AddressBook: React.FC<Props> = ({
  users,
  maxPage,
  isFetching,
  isError,
  getUsers,
  displayedPage,
  displayPage,
  countries
}: Props) => {
  const [modalIsOpen, setIsOpen] = React.useState(false);
  const [selectedUser, setSelectedUser] = React.useState<User | null>(null);
  const [filter, setFilter] = React.useState("");

  useEffect(() => {
    displayPage(true);
    getUsers(countries, true);
  }, [countries]);

  useEffect(() => {
    getUsers(countries);
  }, [users]);

  useBottomScrollListener(() => {
    if (!filter) {
      displayPage();
      getUsers(countries);
    }
  }, 400);

  function openModal() {
    setIsOpen(true);
  }

  function closeModal() {
    setIsOpen(false);
  }

  function handleRowClick(user: User) {
    setSelectedUser(user);
    openModal();
  }

  function handleFilterChange(e: React.ChangeEvent<HTMLInputElement>) {
    setFilter(e.target.value);
  }

  const filterUsers = (user: User) => {
    return `${user.name.first.toLowerCase()} ${user.name.last.toLowerCase()}`.indexOf(
      filter.toLowerCase()
    ) === 0;}

  function displayMessage() {
    if (isError) {
      return "Catalogue unavailable";
    }
    // TODO: check this condition
    if (!isFetching && displayedPage >= maxPage) {
      return "End of user catalogue";
    }

    if (!isFetching && users.length === 0 && displayedPage > 0) {
      return "No users found";
    }
    // TODO: check this condition
    if (
      !filter &&
      !isError &&
      displayedPage < maxPage &&
      (users.length !== 0 || isFetching)
    ) {
      return "Loading...";
    }
  }

  return (
    <div className="address-book">
      <div className="address-book__header">
        <div className="address-book__top-bar">
          <span>Address Book</span>
          <input
            className="address-book__search"
            value={filter}
            onChange={handleFilterChange}
            placeholder="Search..."
          />
          <span className="address-book__settings-link">
            <Link to={paths.settings}>Settings</Link>
          </span>
        </div>
      </div>
      <div className="address-book__list">
        {users.length ? (
          <table className="address-book__table">
            <thead>
              <tr>
                <th>#</th>
                <th></th>
                <th>First Name</th>
                <th>Last Name</th>
                <th>Username</th>
                <th>Email</th>
              </tr>
            </thead>
            <tbody>
              {users
                .slice(0, Math.min(displayedPage, maxPage))
                .reduce((list, page) => list.concat(page), [])
                .filter(filterUsers)
                .map((user, idx) => (
                  <tr key={user.email} onClick={() => handleRowClick(user)}>
                    <td>{idx + 1}</td>
                    <td>
                      <img
                        src={user.picture.thumbnail}
                        width={48}
                        height={48}
                      />
                    </td>
                    <td>{user.name.first}</td>
                    <td>{user.name.last}</td>
                    <td>{user.login.username}</td>
                    <td>{user.email}</td>
                  </tr>
                ))}
            </tbody>
          </table>
        ) : null}
      </div>
      <div className="address-book__message">{displayMessage()}</div>
      {/* TODO: add refresh button */}
      <Modal
        className="address-book__details-modal"
        isOpen={modalIsOpen}
        ariaHideApp={false}
        onRequestClose={closeModal}
        contentLabel="Example Modal"
      >
        {selectedUser && (
          <>
            <p>{"User Details"}</p>
            <div className="address-book__details-container">
              <div>
                <img
                  src={selectedUser.picture.large}
                  width={128}
                  height={128}
                />
              </div>
              <div>
                {`${selectedUser.name.first} ${selectedUser.name.last}`}
                <div>{`${selectedUser.location.street.name} ${selectedUser.location.street.number}`}</div>
                <div>{`${selectedUser.location.city}, ${selectedUser.location.state}`}</div>
                <div>{selectedUser.location.postcode}</div>
                <div>{`pho.: ${selectedUser.phone}`}</div>
                <div>{`cell: ${selectedUser.cell}`}</div>
              </div>
            </div>

            <div className="address-book__details-close">
              <button onClick={closeModal}>close</button>
            </div>
          </>
        )}
      </Modal>
    </div>
  );
};

export default connect(
  (state: State) => ({
    users: state.book.users,
    isFetching: state.book.isFetching,
    isError: state.book.isError,
    countries: state.book.countries,
    displayedPage: state.book.displayedPage
  }),
  { getUsers, displayPage }
)(AddressBook);
