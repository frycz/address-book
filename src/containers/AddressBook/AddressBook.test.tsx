import React from "react";
import { shallow } from "enzyme";
import { AddressBook, Props } from "./AddressBook";
import mockUsers from "../../../mocks/users.json";

const mockProps: Props = {
  users: mockUsers as Props['users'],
  maxPage: 20,
  isFetching: false,
  isError: false,
  countries: { ch: true, es: true, fr: false },
  currentPage: 1,
  getUsers: jest.fn(),
  displayPage: jest.fn()
};

describe("AddressBook", () => {
  it("should match smapshot", () => {
    const settings = shallow(<AddressBook {...mockProps} />);
    expect(settings.debug()).toMatchSnapshot();
  });
});
jest.fn();
