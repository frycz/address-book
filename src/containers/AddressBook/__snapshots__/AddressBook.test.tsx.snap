// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressBook should match smapshot 1`] = `
"<div className=\\"address-book\\">
  <div className=\\"address-book__header\\">
    <div className=\\"address-book__top-bar\\">
      <span>
        Address Book
      </span>
      <input className=\\"address-book__search\\" value=\\"\\" onChange={[Function: handleFilterChange]} placeholder=\\"Search...\\" />
      <span className=\\"address-book__settings-link\\">
        <Link to=\\"/settings\\">
          Settings
        </Link>
      </span>
    </div>
  </div>
  <div className=\\"address-book__list\\">
    <table className=\\"address-book__table\\">
      <thead>
        <tr>
          <th>
            #
          </th>
          <th />
          <th>
            First Name
          </th>
          <th>
            Last Name
          </th>
          <th>
            Username
          </th>
          <th>
            Email
          </th>
        </tr>
      </thead>
      <tbody>
        <tr onClick={[Function: onClick]}>
          <td>
            1
          </td>
          <td>
            <img src=\\"https://randomuser.me/api/portraits/thumb/men/92.jpg\\" width={48} height={48} />
          </td>
          <td>
            Andres
          </td>
          <td>
            Marchand
          </td>
          <td>
            reddog219
          </td>
          <td>
            andres.marchand@example.com
          </td>
        </tr>
        <tr onClick={[Function: onClick]}>
          <td>
            2
          </td>
          <td>
            <img src=\\"https://randomuser.me/api/portraits/thumb/men/65.jpg\\" width={48} height={48} />
          </td>
          <td>
            Jaime
          </td>
          <td>
            Ramirez
          </td>
          <td>
            whitemouse568
          </td>
          <td>
            jaime.ramirez@example.com
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div className=\\"address-book__message\\">
    Loading...
  </div>
  <Modal className=\\"address-book__details-modal\\" isOpen={false} ariaHideApp={false} onRequestClose={[Function: closeModal]} contentLabel=\\"Example Modal\\" portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" role=\\"dialog\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} parentSelector={[Function: parentSelector]} />
</div>"
`;
